alacritty/src/wayland_theme.rs

```diff
@@ -32,51 +35,53 @@ impl AlacrittyWaylandTheme {
         }
     }
 
-    fn color_icon_color(&self, color: Rgb, status: ButtonState) -> [u8; 4] {
-        match status {
-            ButtonState::Hovered => [0xff, color.r, color.g, color.b],
-            ButtonState::Idle => [0xff, self.foreground.r, self.foreground.g, self.foreground.b],
-            ButtonState::Disabled => {
-                [0xff, self.dim_foreground.r, self.dim_foreground.g, self.dim_foreground.b]
-            },
+    fn button_foreground_color(&self, color: Rgb, status: ButtonState, window_active: bool) -> Rgb {
+        match (window_active, status) {
+            (false, _) => self.dim_foreground,
+            (_, ButtonState::Hovered) => color,
+            (_, ButtonState::Idle) => self.foreground,
+            (_, ButtonState::Disabled) => self.dim_foreground,
         }
     }
 }
 
 impl WaylandTheme for AlacrittyWaylandTheme {
-    fn primary_color(&self, _window_active: bool) -> [u8; 4] {
-        [0xff, self.background.r, self.background.g, self.background.b]
-    }
-
-    fn secondary_color(&self, window_active: bool) -> [u8; 4] {
-        if window_active {
-            [0xff, self.foreground.r, self.foreground.g, self.foreground.b]
-        } else {
-            [0xff, self.dim_foreground.r, self.dim_foreground.g, self.dim_foreground.b]
+    fn element_color(&self, element: Element, window_active: bool) -> ARGBColor {
+        match element {
+            Element::Bar | Element::Separator => self.background.into_rgba(),
+            Element::Text if window_active => self.foreground.into_rgba(),
+            Element::Text => self.dim_foreground.into_rgba(),
         }
     }
 
-    fn close_button_color(&self, _status: ButtonState) -> [u8; 4] {
-        [0x00, self.background.r, self.background.g, self.background.b]
-    }
-
-    fn close_button_icon_color(&self, status: ButtonState) -> [u8; 4] {
-        self.color_icon_color(self.hovered_close_icon, status)
-    }
-
-    fn maximize_button_color(&self, _status: ButtonState) -> [u8; 4] {
-        [0x00, self.background.r, self.background.g, self.background.b]
-    }
-
-    fn maximize_button_icon_color(&self, status: ButtonState) -> [u8; 4] {
-        self.color_icon_color(self.hovered_maximize_icon, status)
+    fn button_color(
+        &self,
+        button: Button,
+        state: ButtonState,
+        foreground: bool,
+        window_active: bool,
+    ) -> ARGBColor {
+        match (foreground, button) {
+            (false, _) => ARGBColor { a: 0x00, r: 0x00, g: 0x00, b: 0x00 },
+            (_, Button::Minimize) => self
+                .button_foreground_color(self.hovered_minimize_icon, state, window_active)
```

Comment:
I think it might be way easier to just inline this method? Especially with passing all kinds of data in when none of it is ever used in most branches just gets extremely confusing.

Something like this:

```rust
if !foreground {
    return ARGB::default();
} else if !window_active {
    return self.dim_foreground();
}

match (state, button)  {
    (ButtonState::Idle, _) => self.foreground,
    (ButtonState::Disabled, _) => self.dim_foreground,
    (_, Button::Minimize) => self.hovered_minimize_icon,
    ...
};

```

The last match could also be separated into two I suppose, so you just return early when Idle/Disabled. That would clean up the match a little bit, but would increase the number of early returns which could get a bit messy.